File: Day21/Part3/pal_partition_iv.py
# LC 1745 — Palindrome Partitioning IV
# Check if s can be split into 3 palindromic substrings

def checkPartitioning(s: str) -> bool:
    n = len(s)
    pal = [[False]*n for _ in range(n)]
    for i in range(n-1, -1, -1):
        for j in range(i, n):
            if s[i] == s[j] and (j - i <= 1 or pal[i+1][j-1]):
                pal[i][j] = True
    # Try splits: [0..i], [i+1..j], [j+1..n-1]
    for i in range(n-2):
        if not pal[0][i]:
            continue
        for j in range(i+1, n-1):
            if pal[i+1][j] and pal[j+1][n-1]:
                return True
    return False

if __name__ == "__main__":
    print(checkPartitioning("abcbdd"))  # True

File: learnLeetCode/Day21/Part3/pal_partition_iv.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 1745 — Palindrome Partitioning IV
// Check if string can be split into three palindromic substrings

bool checkPartitioning(string s){
    int n = s.size();
    vector<vector<char>> pal(n, vector<char>(n, 0));
    for(int i=n-1;i>=0;--i){
        for(int j=i;j<n;++j){
            if (s[i]==s[j] && (j-i<=1 || pal[i+1][j-1])) pal[i][j]=1;
        }
    }
    for(int i=0;i<=n-3;++i){
        if (!pal[0][i]) continue;
        for(int j=i+1;j<=n-2;++j){
            if (pal[i+1][j] && pal[j+1][n-1]) return true;
        }
    }
    return false;
}

int main(){
    cout << (checkPartitioning("abcbdd") ? "true" : "false") << "\n"; // true
    return 0;
}
